cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,739 @@
+{
+    return dev->config->device->SYNCBUSY.bit.ENABLE;
+}
+
+static bool _syncbusy_swrst(sam0_common_usb_t *dev)
+{
+    return dev->config->device->SYNCBUSY.bit.SWRST;
+}
+
+static void _gpio_connect_cb(void *arg)
+{
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t*)arg;
+    usbdev->connect_change = true;
+    usbdev->usbdev.cb(&usbdev->usbdev, USBDEV_EVENT_ESR);
+}
+
+static inline void _poweron(void)
+{
+#if defined(CPU_FAM_SAMD21)
+    PM->AHBMASK.reg |= PM_AHBMASK_USB;
+    PM->APBBMASK.reg |= PM_APBBMASK_USB;
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_USB;
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK0;
+#else
+#error "Unknown CPU family for sam0 common usbdev driver"
+#endif
+}
+
+static usbdev_ep_t *_usbdev_new_ep(usbdev_t *dev, usb_ep_type_t type,
+                                  usb_ep_dir_t dir, size_t buf_len)
+{
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t*)dev;
+    /* The IP supports all types for all endpoints */
+    usbdev_ep_t *res = NULL;
+
+    /* Always return endpoint 0 for control types */
+    if (type == USB_EP_TYPE_CONTROL) {
+        res = _get_ep(usbdev, 0, dir);
+        res->num = 0;
+    }
+    else {
+        /* Find the first unassigned ep with proper dir */
+        for (unsigned idx = 1; idx < SAM_USB_NUM_EP && !res; idx++) {
+            usbdev_ep_t *ep = _get_ep(usbdev, idx, dir);
+            if (ep->type == USB_EP_TYPE_NONE) {
+                res = ep;
+                res->num = idx;
+            }
+        }
+    }
+    if (res) {
+        res->dev = dev;
+        res->dir = dir;
+        if (usbdev->used + buf_len < SAM_USB_BUF_SPACE) {
+            res->buf = usbdev->buffer + usbdev->used;
+            res->len = buf_len;
+            if (_bank_set_size(res) < 0) {
+                return NULL;
+            }
+            usbdev->used += buf_len;
+            _bank_set_address(res);
+            res->type = type;
+            res->dev = dev;
+        }
+    }
+    return res;
+}
+
+void sam_usbdev_setup(sam0_common_usb_t *usbdev,
+                      const sam0_common_usb_config_t *config)
+{
+    usbdev->usbdev.driver = &driver;
+    usbdev->config = config;
+    devs[0] = usbdev;
+}
+
+static void _usbdev_init(usbdev_t *dev)
+{
+    DEBUG("Initializing sam0 usb peripheral\n");
+    /* Only one usb device on this board */
+    sam0_common_usb_t * usbdev = (sam0_common_usb_t *)dev;
+    usbdev->used = 0;
+    /* Start disconnected from the host */
+    usbdev->connect_change = false;
+    /* Set GPIO */
+    gpio_init(GPIO_PIN(PA, 24), GPIO_IN);
+    gpio_init(GPIO_PIN(PA, 25), GPIO_IN);
+    gpio_init_mux(GPIO_PIN(PA, 24), GPIO_MUX_G);
+    gpio_init_mux(GPIO_PIN(PA, 25), GPIO_MUX_G);
```

Comment:
You should reused your config struct defined in periph_conf.h here.

Reply:
So much effort spent on the config struct, only to forget to use it in the implementation…